### PR TITLE
Various fixes to allow Java PAssert to run on Python

### PR DESCRIPTION
**Pull request by robertwb**
* Fixes to length prefixing.
* Never trigger
* Unknown non-merging window support.
* Use Flatten as identity DoFn for transcoding.
* Fix GetState to also return timestamp.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @kennknowles

File: sdks/python/apache_beam/transforms/trigger_test.py [hunk 1]
```diff
@@ -518,6 +519,28 @@ def format_result(k_v):
...
+          | beam.Create([1, 1, 2, 3, 4, 5, 10, 11])
+          | beam.FlatMap(lambda t: [('A', t), ('B', t + 5)])
+          | beam.Map(construct_timestamped)
+          | beam.WindowInto(
+              FixedWindows(10),
+              trigger=Never(),
+              accumulation_mode=AccumulationMode.DISCARDING)
+          | beam.GroupByKey()
+          | beam.Map(format_result))
+      assert_that(result, equal_to([]))
```
Comment (kennknowles) [inline sdks/python/apache_beam/transforms/trigger_test.py:542]: Never trigger is misnamed. It means that the trigger itself never fires, but that at window GC output is produced. That's why PAssert uses it to gather the full result.

Comment (robertwb) [reply sdks/python/apache_beam/transforms/trigger_test.py:542]: Ack. (Interestingly, I originally was expecting it to fire at EoW.) The ULR doesn't yet support gc timers. I can look into fixing this (or at least making the "Never" trigger correct). 

As an aside, the non-trivial triggering and windowing in PAssert makes it less than ideal for validating new runners.

Comment (robertwb) [reply sdks/python/apache_beam/transforms/trigger_test.py]: I couldn't find an existing bug about closing behaviors, so I filed BEAM-10149. That seems pretty invasive to change as part of this PR, so I added a hack to support its use with proper closing behavior in batch only. PTAL.

Comment (kennknowles) [reply sdks/python/apache_beam/transforms/trigger_test.py]: Good point about that in PAssert. Though isn't the triggering just "fire at GC" and the windowing nothing special? Previously, it depended on side inputs which tend to come later in a runner's lifecycle.

Comment (robertwb) [reply sdks/python/apache_beam/transforms/trigger_test.py]: I think even the concept of GC triggers (which have to be scheduled at a different time than a "normal" trigger at EOW) and late data is rather advanced. Also, yes, it used a custom (java serializable) WindowFn (didn't investigate which one this was). Possibly the right thing to do here would be to use "standard" triggers for the normal case, and advanced features when needed, assuming they can be queried at pipeline construction time. But this is completely an aside. 

Thanks for the review. I discovered another couple of issues when running the full suite; I'll get them into separate PRs.

**Comment by robertwb** [comment]
Run Python PreCommit

**Comment by robertwb** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by robertwb** [comment]
Run Python PreCommit

**Comment by robertwb** [comment]
The apache_beam.io.gcp.bigquery_file_loads_test.TestBigQueryFileLoads.test_multiple_partition_files
 failures seem irrelevant.

**Comment by robertwb** [comment]
Run Python PreCommit

**Comment by robertwb** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by robertwb** [comment]
Run Python PreCommit

**Pull request merged**